daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala

```diff
@@ -92,9 +96,161 @@ object TransactionConversions {
       .sequence_
       .map(_ => RawTransaction(transactionBuilder.build.toByteString))
   }
+
+  /** Decodes and extracts outputs of a submitted transaction, that is the IDs and keys of contracts created or updated
+    * by processing a submission.
+    *
+    * The results, among the others, may include contract IDs of transient contracts (created and archived within the same transaction),
+    * contract IDs that may cause divulgence and keys that are not modified. Actual outputs must be a subset of,
+    * or the same as, computed outputs and we currently relax this check by widening the latter set,
+    * treating a node the same regardless of whether it was a child of a rollback node or not, for example.
+    * Computed outputs that are not actual outputs can be safely trimmed.
```

Comment:
Reading the comments by @cocreature (thanks for the detailed review!) and the commentary here, I believe it would be valuable to make an attempt to clarify further what exactly the "transaction output" are that are computed by this function. The comment as-is helps only partially as it references several concepts that are undefined a this dependency level of code; e.g. "actual outputs", "computed outputs", "safely trimmed", and likely also "processing a submission".

AFAIS, one viable definitions would be that that the outputs of a transaction `tx` are exactly (1) the contract-ids of contracts that `tx` archives or divulges and (2) the keys whose "key state" (as per https://docs.daml.com/concepts/ledger-model/ledger-integrity.html) is changed by `tx`.

There are other definitions possible, but I believe we should aim for a clean definition built on top of the ledger model here to avoid miscommunication between the language team and the kv team.

What do you think @hubert-da @miklos-da @cocreature @remyhaemmerle-da ? Can we find such a definition for the semantics required by the kv team?

Reply:
Definitely in favor of cleaning this up and we were planning to do that as a follow-up. This PR was only reshuffling code so I just wanted to document the current semantics rather than changing them.

The semantics you suggested mostly sound fine to me but unless I’m missing something, I think we should exclude contracts created in the current transaction that are not active at the end. So slightly rephrased:

1. A contract id is an output if  one of the following is true:
      1. The transaction contains a fetch (potentially by key) or exercise for that contract id and the contract was not created in the current transaction. This captures divulgence and includes archives. Note that a lookup by key does not divulge (atm, I recall us wanting to change that but afaict we have not done so yet) so I’d not include the contract id for a successful lookup by key.
      2. The transaction contains a create for that contract id and the contract is still active at the end of the transaction. This excludes transient contracts and rolled back creates.
 2. Your key state definition sounds fine for keys.

Reply:
Actually, after doing some further code spelunking, it looks to me that this is the only real use-site for the `extractTransactionOutputs` function: https://github.com/digital-asset/daml/blob/9ed1175d41237262fd195f7519ff512bf64e32c9/ledger/participant-state/kvutils/src/test/lib/scala/com/daml/ledger/participant/state/kvutils/KVTest.scala#L388

It is not clear to me why the check needs to be done against the result of `extractTransactionOutputs`. I'd suggest to change the usesite to check that
1. the read-set of the execution result is a subset of the inputs that were declared in the submission
2. the outputs are a subset of the read-set, which ensures that there is no write to a kv-key whose value has not been read. 

This way we can remove the need for an extra function and make the test more precise.

What do you think @hubert-da and @miklos-da ?

Reply:
Thanks! I'll try to clean this up.

Reply:
@meiersi-da, the `extractTransactionOutputs`/`transactionOutputs` function is also being used by `KeyValueCommitting.verifyStateUpdatesAgainstPreDeclaredOutputs` that on the other hand is being used by `KeyValueCommitting.processSubmission`.

@miklos-da, is there anything we'd like to do with that? Based on your comment and the above, I don't think we do.